packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js

```diff
@@ -0,0 +1,124 @@
+/**
+ * WordPress dependencies
+ */
+import { useRegistry } from '@wordpress/data';
+import { store as coreStore } from '@wordpress/core-data';
+import {
+	useReducer,
+	useCallback,
+	useRef,
+	useLayoutEffect,
+} from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import useCreateNavigationMenu from './use-create-navigation-menu';
+import menuItemsToBlocks from '../menu-items-to-blocks';
+
+function reducer( state, action ) {
+	switch ( action.type ) {
+		case 'RESOLVED':
+			return {
+				...state,
+				isFetching: false,
+				navMenu: action.navMenu,
+			};
+		case 'ERROR':
+			return {
+				...state,
+				isFetching: false,
+				navMenu: null,
+			};
+		case 'LOADING':
+			return {
+				...state,
+				isFetching: true,
+			};
+		default:
+			throw new Error( `Unexpected action type ${ action.type }` );
+	}
+}
+
+function useSafeDispatch( dispatch ) {
+	const mounted = useRef( false );
+	useLayoutEffect( () => {
+		mounted.current = true;
+		return () => ( mounted.current = false );
+	}, [] );
+	return useCallback(
+		( ...args ) => ( mounted.current ? dispatch( ...args ) : void 0 ),
+		[ dispatch ]
+	);
+}
+
+function useConvertClassicToBlockMenu( clientId ) {
+	const createNavigationMenu = useCreateNavigationMenu( clientId );
+	const registry = useRegistry();
+
+	const [ state, dispatch ] = useReducer( reducer, {
+		navMenu: null,
+		isFetching: false,
+	} );
+
+	const safeDispatch = useSafeDispatch( dispatch );
+
+	async function convertClassicMenuToBlockMenu( menuId, menuName ) {
```

Comment:
I'd pull it out of the hook, there's no need to construct a new function on each run.

Reply:
I take your point but if you follow the flow you'll notice that this is required to be within a hook because it uses other hooks. In fact it's hooks within hooks within hooks.

I'm not in favour of this approach really but I don't think we should look to refactor in this PR because it will become overly complex to review.

How would you feel about pairing on a refactor of this sometime?

Reply:
Sure, let's try to pair up tomorrow

Reply:
With the recent updates in place I feel pretty good about having this here, let's just wrap it in `useCallback`.